doc/source/serve/doc_code/vllm_openai_example.py

```diff
@@ -0,0 +1,115 @@
+from vllm.entrypoints.openai.serving_chat import OpenAIServingChat
+from vllm.engine.arg_utils import AsyncEngineArgs
+from vllm.engine.async_llm_engine import AsyncLLMEngine
+from vllm.entrypoints.openai.protocol import ChatCompletionRequest
+from vllm.entrypoints.openai.protocol import ChatCompletionResponse
+from vllm.entrypoints.openai.serving_chat import OpenAIServingChat
+from vllm.entrypoints.openai.serving_engine import LoRAModulePath
+from vllm.entrypoints.openai.cli_args import make_arg_parser
+
+logger = logging.getLogger("ray.serve")
+
+app = FastAPI()
+
+@serve.deployment(num_replicas=2,
+                  max_concurrent_queries=10)
+@serve.ingress(app)
+class VLLMDeployment:
+    def __init__(self,
+            engine_args: AsyncEngineArgs,
+            response_role: str,
+            lora_modules: Optional[List[LoRAModulePath]]=None,
+            chat_template: Optional[str] = None):
+
+        logger.info(engine_args)
+        self.engine = AsyncLLMEngine.from_engine_args(engine_args)
+        if engine_args.served_model_name is not None:
+            served_model_names = engine_args.served_model_name
+        else:
+            served_model_names = [engine_args.model]
+        self.openai_serving_chat = OpenAIServingChat(self.engine,
+                                                        served_model_names,
+                                                        response_role,
+                                                        lora_modules,
+                                                        chat_template)
+
+    # API Reference here - https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html
+    @app.post("/v1/chat/completions")
+    async def create_chat_completion(self,
+                                request: ChatCompletionRequest,
+                                raw_request: Request):
+        print("request", request)
+        print("raw request", raw_request)
+        generator = await self.openai_serving_chat.create_chat_completion(
+            request, raw_request)
+        if isinstance(generator, ErrorResponse):
+            return JSONResponse(content=generator.model_dump(),
+                                status_code=generator.code)
+        if request.stream:
+            return StreamingResponse(content=generator,
+                                    media_type="text/event-stream")
+        else:
+            assert isinstance(generator, ChatCompletionResponse)
+            return JSONResponse(content=generator.model_dump())
+
+def deployment(cli_args: Dict[str, str]) -> serve.Application:
+    """
+    Refer to https://docs.vllm.ai/en/latest/models/engine_args.html for arguments
+
+    """
+    parser = make_arg_parser()
+    arg_strings = []
+    for key, value in cli_args.items():
+        arg_strings.extend([f'--{key}', str(value)])
+    print(arg_strings)
+    parsed_args = parser.parse_args(args=arg_strings)
+    engine_args = AsyncEngineArgs.from_cli_args(parsed_args)
```

Comment:
this is pretty janky.. let's define a pydantic model for the args instead (recommended pattern)

Reply:
then you can leave comments explaining what the args are too